File: java/src/pushy/internal/Message.java
/*
 * Copyright (c) 2009 Andrew Wilkins <axwalk@gmail.com>
 *
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated documentation
 * files (the "Software"), to deal in the Software without
 * restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following
 * conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */

package pushy.internal;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * A class for decoding, encoding and describing a message.
 */
public class Message
{
    private final Type   type;
    private final byte[] payload;
    private long         target;
    private long         source;

    /**
     * Create a new message, with the current thread designated as source
     * of the message.
     */
    public Message(Type type, byte[] payload, long target)
    {
        this(type, payload, target, ThreadId.getThreadId());
    }

    private Message(Type type, byte[] payload, long target, long source)
    {
        this.type    = type;
        this.payload = payload;
        this.target  = target;
        this.source  = source;
    }

    public Type getType()
    {
        return type;
    }

    public long getSource()
    {
        return source;
    }

    public long getTarget()
    {
        return target;
    }

    public final byte[] getPayload()
    {
        return payload;
    }

    public String toString()
    {
        return "Message(" + type + ", " + getSource() + ", " +
                getTarget() + ", " + payload.length + " bytes)";
    }

    public boolean equals(Object rhs)
    {
        if (rhs instanceof Message)
        {
            Message other = (Message)rhs;
            return type.equals(other.type) &&
                   target == other.target &&
                   source == other.source &&
                   Arrays.equals(payload, other.payload);
        }
        return super.equals(rhs);
    }

    /**
     * Pack a message into its network representation.
     */
    public byte[] pack()
    {
        // 21 bytes for the header + the payload.
        java.io.ByteArrayOutputStream stream =
            new java.io.ByteArrayOutputStream(21 + payload.length);
        try {
            pack(stream);
        } catch (java.io.IOException e) {}
        return stream.toByteArray();
    }

    /**
     * Pack a message into its network representation.
     */
    public void pack(java.io.OutputStream stream) throws java.io.IOException
    {
        stream.write((byte)type.getCode());
        pack(stream, source);
        pack(stream, target);
        pack(stream, payload.length);
        stream.write(payload, 0, payload.length);
    }

    /**
     * Read a message from the given input stream.
     */
    public static Message
    unpack(java.io.InputStream stream) throws java.io.IOException
    {
        // Unpack the header.
        Type type = Type.getType(stream.read());
        long source = unpackLong(stream);
        long target = unpackLong(stream);
        int length = unpackInteger(stream);

        // Read the payload and create the message.
        byte[] payload = read(stream, length);
        return new Message(type, payload, target, source);
    }

    // Utility method for creating and filling up a byte array from an
    // input stream.
    private static byte[]
    read(java.io.InputStream stream, int length) throws java.io.IOException
    {
        byte[] buf = new byte[length];
        read(stream, buf);
        return buf;
    }

    // Utility method for filling up a byte array from an input stream.
    private static void
    read(java.io.InputStream stream, byte[] buf) throws java.io.IOException
    {
        int nread = 0;
        do
        {
            int partial = stream.read(buf, nread, buf.length-nread);
            if (partial == -1)
                throw new java.io.EOFException();
            nread += partial;
        } while (nread < buf.length);
    }

    // Utility method for packing a network-order "long" into a byte array.
    private static void
    pack(java.io.OutputStream stream, long value) throws java.io.IOException
    {
        stream.write((byte)((value >> 56) & 0xFF));
        stream.write((byte)((value >> 48) & 0xFF));
        stream.write((byte)((value >> 40) & 0xFF));
        stream.write((byte)((value >> 32) & 0xFF));
        stream.write((byte)((value >> 24) & 0xFF));
        stream.write((byte)((value >> 16) & 0xFF));
        stream.write((byte)((value >> 8) & 0xFF));
        stream.write((byte)((value) & 0xFF));
    }

    // Utility method for unpacking a network-order "long" from a byte array.
    private static long
    unpackLong(java.io.InputStream stream) throws java.io.IOException
    {
        long hi4 = (long)unpackInteger(stream);
        long lo4 = (long)unpackInteger(stream);
        return (hi4 << 4) | lo4;
    }

    // Utility method for packing a network-order "int" into a byte array.
    private static void
    pack(java.io.OutputStream stream, int value) throws java.io.IOException
    {
        stream.write((byte)((value >> 24) & 0xFF));
        stream.write((byte)((value >> 16) & 0xFF));
        stream.write((byte)((value >> 8) & 0xFF));
        stream.write((byte)((value) & 0xFF));
    }

    // Utility method for unpacking a network-order "int" from a byte array.
    private static int
    unpackInteger(java.io.InputStream stream) throws java.io.IOException
    {
        int b0 = stream.read();
        int b1 = stream.read();
        int b2 = stream.read();
        int b3 = stream.read();
        if (b0 == -1 || b1 == -1 || b2 == -1 || b3 == -1)
            throw new java.io.EOFException();
        return (b0 << 24) | (b1 << 16) | (b2 << 8) | b3;
    }

    /**
     * A class for describing a message type.
     */
    public static class Type
    {
        private int code;
        private String name;

        public Type(int code, String name)
        {
            this.code = code;
            this.name = name;
        }

        public int getCode()
        {
            return code;
        }

        public String getName()
        {
            return name;
        }

        public String toString()
        {
            return "MessageType(" + code + ", '" + name + "')";
        }

        public int hashCode()
        {
            return code;
        }

        public boolean equals(Object rhs)
        {
            if (rhs instanceof Type)
                return code == ((Type)rhs).code;
            else if (rhs instanceof Integer)
                return code == ((Integer)rhs).intValue();
            else if (rhs instanceof String)
                return name.equals(rhs);
            return super.equals(rhs);
        }

        // Provide a means of getting a message type by its code.
        private static List types = new ArrayList();
        public static Type getType(int code)
        {
            if (code < 0 || code >= types.size())
                return null;
            return (Type)types.get(code);
        }

        public static List getTypes()
        {
            return java.util.Collections.unmodifiableList(types);
        }

        // Method for defining a type, given a name. The type's code
        // will be the next index into the 'types' list.
        private static Type createType(String name)
        {
            Type type = new Type(types.size(), name);
            types.add(type);
            return type;
        }

        // Define message types: must be in the same order as in the Python
        // code.
        public static final Type evaluate       = createType("evaluate");
        public static final Type response       = createType("response");
        public static final Type exception      = createType("exception");
        public static final Type getattr        = createType("getattr");
        public static final Type setattr        = createType("setattr");
        public static final Type getstr         = createType("getstr");
        public static final Type getrepr        = createType("getrepr");
        public static final Type op__call__     = createType("op__call__");
        public static final Type op__lt__       = createType("op__lt__");
        public static final Type op__le__       = createType("op__le__");
        public static final Type op__eq__       = createType("op__eq__");
        public static final Type op__ne__       = createType("op__ne__");
        public static final Type op__gt__       = createType("op__gt__");
        public static final Type op__ge__       = createType("op__ge__");
        public static final Type op__cmp__      = createType("op__cmp__");
        public static final Type op__rcmp__     = createType("op__rcmp__");
        public static final Type op__hash__     = createType("op__hash__");
        public static final Type op__nonzero__  = createType("op__nonzero__");
        public static final Type op__unicode__  = createType("op__unicode__");
        public static final Type op__len__      = createType("op__len__");
        public static final Type op__getitem__  = createType("op__getitem__");
        public static final Type op__setitem__  = createType("op__setitem__");
        public static final Type op__delitem__  = createType("op__delitem__");
        public static final Type op__iter__     = createType("op__iter__");
        public static final Type op__contains__ = createType("op__contains__");
        public static final Type op__get__      = createType("op__get__");
        public static final Type op__set__      = createType("op__set__");
        public static final Type op__delete__   = createType("op__delete__");
        public static final Type op__getslice__ = createType("op__getslice__");
        public static final Type op__setslice__ = createType("op__setslice__");
        public static final Type op__delslice__ = createType("op__delslice__");
        public static final Type op__add__      = createType("op__add__");
        public static final Type op__sub__      = createType("op__sub__");
        public static final Type op__mul__      = createType("op__mul__");
        public static final Type op__floordiv__ = createType("op__floordiv__");
        public static final Type op__mod__      = createType("op__mod__");
        public static final Type op__divmod__   = createType("op__divmod__");
        public static final Type op__pow__      = createType("op__pow__");
        public static final Type op__lshift__   = createType("op__lshift__");
        public static final Type op__rshift__   = createType("op__rshift__");
        public static final Type op__and__      = createType("op__and__");
        public static final Type op__xor__      = createType("op__xor__");
        public static final Type op__or__       = createType("op__or__");
        public static final Type op__div__      = createType("op__div__");
        public static final Type op__truediv__  = createType("op__truediv__");
        public static final Type op__radd__     = createType("op__radd__");
        public static final Type op__rsub__     = createType("op__rsub__");
        public static final Type op__rdiv__     = createType("op__rdiv__");
        public static final Type op__rtruediv__ = createType("op__rtruediv__");
        public static final Type op__rfloordiv__
            = createType("op__rfloordiv__");
        public static final Type op__rmod__     = createType("op__rmod__");
        public static final Type op__rdivmod__  = createType("op__rdivmod__");
        public static final Type op__rpow__     = createType("op__rpow__");
        public static final Type op__rlshift__  = createType("op__rlshift__");
        public static final Type op__rrshift__  = createType("op__rrshift__");
        public static final Type op__rand__     = createType("op__rand__");
        public static final Type op__rxor__     = createType("op__rxor__");
        public static final Type op__ror__      = createType("op__ror__");
        public static final Type op__iadd__     = createType("op__iadd__");
        public static final Type op__isub__     = createType("op__isub__");
        public static final Type op__imul__     = createType("op__imul__");
        public static final Type op__idiv__     = createType("op__idiv__");
        public static final Type op__itruediv__ = createType("op__itruediv__");
        public static final Type op__ifloordiv__
            = createType("op__ifloordiv__");
        public static final Type op__imod__     = createType("op__imod__");
        public static final Type op__ipow__     = createType("op__ipow__");
        public static final Type op__ilshift__  = createType("op__ilshift__");
        public static final Type op__irshift__  = createType("op__irshift__");
        public static final Type op__iand__     = createType("op__iand__");
        public static final Type op__ixor__     = createType("op__ixor__");
        public static final Type op__ior__      = createType("op__ior__");
        public static final Type op__neg__      = createType("op__neg__");
        public static final Type op__pos__      = createType("op__pos__");
        public static final Type op__abs__      = createType("op__abs__");
        public static final Type op__invert__   = createType("op__invert__");
        public static final Type op__complex__  = createType("op__complex__");
        public static final Type op__int__      = createType("op__int__");
        public static final Type op__long__     = createType("op__long__");
        public static final Type op__float__    = createType("op__float__");
        public static final Type op__oct__      = createType("op__oct__");
        public static final Type op__hex__      = createType("op__hex__");
        public static final Type op__index__    = createType("op__index__");
        public static final Type op__coerce__   = createType("op__coerce__");
        public static final Type op__enter__    = createType("op__enter__");
        public static final Type op__exit__     = createType("op__exit__");

        /**
         * Check if a message type is a a response type.
         */
        public boolean isResponse()
        {
            return this.equals(response) || this.equals(exception);
        }
    }
}

File: pushy/protocol/message.py
# Copyright (c) 2008, 2009 Andrew Wilkins <axwalk@gmail.com>
# 
# Permission is hereby granted, free of charge, to any person
# obtaining a copy of this software and associated documentation
# files (the "Software"), to deal in the Software without
# restriction, including without limitation the rights to use,
# copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following
# conditions:
# 
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
# OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
# HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
# WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
# OTHER DEALINGS IN THE SOFTWARE.

import os, struct, thread
import pushy.util

class MessageType:
    "A class for describing the type of a message."

    def __init__(self, code, name):
        self.code = code
        self.name = name
    def __repr__(self):
        return "MessageType(%d, '%s')" % (self.code, self.name)
    def __int__(self):
        return self.code
    def __str__(self):
        return self.name
    def __hash__(self):
        return self.code
    def __eq__(self, other):
        if type(other) is int: return other == self.code
        elif type(other) is str: return other == self.name
        return other.__class__ is MessageType and other.code == self.code


class Message:
    PACKING_FORMAT = ">BqqI"
    PACKING_SIZE   = struct.calcsize(PACKING_FORMAT)

    def __init__(self, type, payload, target=0, source=None):
        self.type     = type
        self.payload  = payload
        self.target   = target
        if source is None:
            source = thread.get_ident()
        self.source = source

    def __eq__(self, other):
        if other.__class__ is not Message: return False
        return self.type == other.type and \
               self.source == other.source and \
               self.target == other.target and \
               self.payload == other.payload

    def __repr__(self):
        return "Message(%r, %d->%d, %r [%d bytes])" % \
                   (self.type, self.source, self.target, self.payload,
                    len(self.payload))

    def pack(self):
        result = struct.pack(self.PACKING_FORMAT, int(self.type), self.source,
                             self.target, len(self.payload))
        return result + self.payload

    @staticmethod
    def unpack(file):
        # Read the message header.
        data = ""
        while len(data) < Message.PACKING_SIZE:
            try:
                partial = file.read(Message.PACKING_SIZE - len(data))
            except Exception, e:
                raise IOError, e
            if partial == "":
                raise IOError, "End of file"
            data += partial
        (type, source, target, length) = \
            struct.unpack(Message.PACKING_FORMAT, data)
        type = message_types[type]

        # Read message payload.
        payload = ""
        if length:
            while len(payload) < length:
                try:
                    partial = file.read(length - len(payload))
                except Exception, e:
                    raise IOError, e
                if partial == "":
                    raise IOError, "End of file"
                payload += partial

        pushy.util.logger.debug(
            "Read %d, %r %r", len(data) + len(payload),
            data + payload, (type, source, target, length))
        return Message(type, payload, target, source)

###############################################################################
# Create enumeration of message types
###############################################################################

message_names = (
  "evaluate",
  "response",
  "exception",
  "getattr",
  "setattr",
  "getstr",
  "getrepr",

  # All object ops go at the end
  "op__call__",
  "op__lt__",
  "op__le__",
  "op__eq__",
  "op__ne__",
  "op__gt__",
  "op__ge__",
  "op__cmp__",
  "op__rcmp__",
  "op__hash__",
  "op__nonzero__",
  "op__unicode__",
  "op__len__",
  "op__getitem__",
  "op__setitem__",
  "op__delitem__",
  "op__iter__",
  "op__contains__",
  #"op__slots__",
  "op__get__",
  "op__set__",
  "op__delete__",
  "op__getslice__",
  "op__setslice__",
  "op__delslice__",
  "op__add__",
  "op__sub__",
  "op__mul__",
  "op__floordiv__",
  "op__mod__",
  "op__divmod__",
  "op__pow__",
  "op__lshift__",
  "op__rshift__",
  "op__and__",
  "op__xor__",
  "op__or__",
  "op__div__",
  "op__truediv__",
  "op__radd__",
  "op__rsub__",
  "op__rdiv__",
  "op__rtruediv__",
  "op__rfloordiv__",
  "op__rmod__",
  "op__rdivmod__",
  "op__rpow__",
  "op__rlshift__",
  "op__rrshift__",
  "op__rand__",
  "op__rxor__",
  "op__ror__",
  "op__iadd__",
  "op__isub__",
  "op__imul__",
  "op__idiv__",
  "op__itruediv__",
  "op__ifloordiv__",
  "op__imod__",
  "op__ipow__",
  "op__ilshift__",
  "op__irshift__",
  "op__iand__",
  "op__ixor__",
  "op__ior__",
  "op__neg__",
  "op__pos__",
  "op__abs__",
  "op__invert__",
  "op__complex__",
  "op__int__",
  "op__long__",
  "op__float__",
  "op__oct__",
  "op__hex__",
  "op__index__",
  "op__coerce__",
  "op__enter__",
  "op__exit__",
)
message_types = []
for i,t in enumerate(message_names):
    m = MessageType(i, t)
    message_types.append(m)
    setattr(MessageType, t, m)
message_types = tuple(message_types)

